### PR TITLE
Add GCP PubSub output feature

### PR DESCRIPTION
**Pull request by IanRobertson-wpe**
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area outputs

**What this PR does / why we need it**:

This adds a PubSub output for users of Google Cloud Platform.

**Which issue(s) this PR fixes**:

None

**Special notes for your reviewer**:

This has been tested in a GKE environment on 8 varying clusters with 50+ nodes.  100% 'ok' observed through /debug/vars endpoint with over 90k alerts generated.

### CODE REVIEW

**Comment by Issif** [comment]
Awesome. Thanks @IanRobertson-wpe.

Can you sign-off your commits please, our CI will refuse to merge without.

For my part I'll review and comment your changes.

**Comment by Issif** [comment]
I think you should rebase on upstream and fix conflicts too, please.

File: README.md [hunk 1]
```diff
@@ -224,6 +225,12 @@ discord:
   # icon: "" # Discord icon (avatar)
   # minimumpriority: "debug" # minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
 
+gcp:
+  credentials: "" # The base64-encoded JSON key file for the GCP service account
+  pubsub:
+    projectid: "" # The GCP Project ID containing the Pub/Sub Topic
+    topic: "" # The name of the Pub/Sub topic
+
```
Comment (Issif) [inline README.md:233]:
```suggestion
  # minimumpriority: "debug" # minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
```

File: config.go [hunk 2]
```diff
@@ -103,6 +103,9 @@ func getConfig() *types.Configuration {
 	v.SetDefault("Azure.eventHub.Namespace", "")
 	v.SetDefault("Azure.eventHub.Name", "")
 	v.SetDefault("Azure.eventHub.MinimumPriority", "")
+	v.SetDefault("GCP.Credentials", "")
+	v.SetDefault("GCP.PubSub.ProjectID", "")
+	v.SetDefault("GCP.PubSub.Topic", "")
```
Comment (Issif) [inline config.go:108]:
```suggestion
	v.SetDefault("GCP.PubSub.Topic", "")
	v.SetDefault("GCP.PubSub.MinimumPriority", "")
```

File: config.go [hunk 3]
```diff
@@ -151,6 +154,7 @@ func getConfig() *types.Configuration {
 	c.Opsgenie.MinimumPriority = checkPriority(c.Opsgenie.MinimumPriority)
 	c.Webhook.MinimumPriority = checkPriority(c.Webhook.MinimumPriority)
 	c.Azure.EventHub.MinimumPriority = checkPriority(c.Azure.EventHub.MinimumPriority)
+	c.GCP.PubSub.MinimumPriority = checkPriority(c.Azure.EventHub.MinimumPriority)
```
Comment (Issif) [inline config.go:157]:
```suggestion
	c.GCP.PubSub.MinimumPriority = checkPriority(c.PubSub.EventHub.MinimumPriority)
```

File: outputs/gcp.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	"errors"
+	"encoding/json"
+	"github.com/DataDog/datadog-go/statsd"
+	"github.com/falcosecurity/falcosidekick/types"
+)
+
+// NewGCPClient returns a new output.Client for accessing the GCP API.
+func NewGCPClient(config *types.Configuration, stats *types.Statistics, statsdClient, dogstatsdClient *statsd.Client) (*Client, error) {
+
+	if config.GCP.Credentials != "" {
```
Comment (Issif) [inline outputs/gcp.go:20]: why do you set env vars?

File: outputs/gcp.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+		}
+		pubSubClient, err := pubsub.NewClient(context.Background(), config.GCP.PubSub.ProjectID, option.WithCredentials(credentials))
+		if err != nil {
+			log.Printf("[ERROR] : GCP PubSub - %v\n", "Error while creating GCP PubSub Client")
+			return nil, errors.New("Error while creating GCP PubSub Client")
+		}
+		topicClient = pubSubClient.Topic(config.GCP.PubSub.Topic)
+	}
+	return &Client{
+		OutputType:      "GCP",
```
Comment (Issif) [inline outputs/gcp.go:48]:
```suggestion
		OutputType:      "GCPPubSub",
```

**Review by Issif — CHANGES_REQUESTED** [review]

**Comment by Issif** [comment]
After your rebase, could you integrate `prometheus` metrics for your Output too? See: https://github.com/falcosecurity/falcosidekick/blob/d35ff2b22b1fe6765981d4a8c13136a882b9caec/outputs/slack.go#L136

File: stats.go [hunk 6]
```diff
@@ -21,26 +21,27 @@ func getInitStats() *types.Statistics {
...
+		AWSLambda:     expvar.NewMap("outputs.awslambda"),
+		AWSSQS:        expvar.NewMap("outputs.awssqs"),
+		AWSSNS:        expvar.NewMap("outputs.awssns"),
+		SMTP:          expvar.NewMap("outputs.smtp"),
+		Opsgenie:      expvar.NewMap("outputs.opsgenie"),
+		Statsd:        expvar.NewMap("outputs.statsd"),
+		Dogstatsd:     expvar.NewMap("outputs.dogstatsd"),
+		Webhook:       expvar.NewMap("outputs.webhook"),
+		AzureEventHub: expvar.NewMap("outputs.azureeventhub"),
+		GCPPubSub:     expvar.NewMap("outputs.gcppubsub"),
```
Comment (Issif) [inline stats.go:44]: this whole block has been changed in last commit by a new method for declaring expvar, please rebase on upstream master

File: stats.go [hunk 7]
```diff
@@ -51,7 +52,69 @@ func getInitStats() *types.Statistics {
...
+	stats.Dogstatsd.Add("ok", 0)
+	stats.Webhook.Add("total", 0)
+	stats.Webhook.Add("error", 0)
+	stats.Webhook.Add("ok", 0)
+	stats.AzureEventHub.Add("total", 0)
+	stats.AzureEventHub.Add("error", 0)
+	stats.AzureEventHub.Add("ok", 0)
+        stats.GCPPubSub.Add("total", 0)
+        stats.GCPPubSub.Add("error", 0)
+        stats.GCPPubSub.Add("ok", 0)
```
Comment (Issif) [inline stats.go:117]: wrong identation, and see my preivous comment, this type of declaration is now deprecated

**Review by Issif — CHANGES_REQUESTED** [review]

File: outputs/gcp.go (hunk 5 above)
Comment (IanRobertson-wpe) [reply outputs/gcp.go:40]: I followed the example of the AWS output type, which is "AWS". I think this is more appropriate here as the intention is to allow for other GCP output types in the future, hence the more generic form of "GCP". Happy to change if you disagree.

File: outputs/gcp.go (hunk 4 above)
Comment (IanRobertson-wpe) [reply outputs/gcp.go]: I had followed the style in the AWS output, but I thought it was odd. I've removed this.

File: stats.go (hunk 6 above)
Comment (IanRobertson-wpe) [reply stats.go]: This has been corrected.

File: outputs/gcp.go (hunk 5 above)
Comment (Issif) [reply outputs/gcp.go:40]: You're right, thanks.

**Review by Issif — APPROVED** [review]

**Comment by Issif** [comment]
/lgtm

👍 @IanRobertson-wpe thanks a lot, I will release 2.17.0 asap for sure

**Pull request merged**